File: integrations/old-source-preservation/ivault-eye-tracking/snapshot/src/features/remote/remoteRouter.ts
import {
  buildConfirmationCopy,
  evaluateRemotePermission,
  isHighRiskCommandType,
} from "./remotePolicy";
import type {
  RemoteCommand,
  RemoteCommandHandlerMap,
  RemoteCommandLogItem,
  RemoteCommandType,
  RemoteControlState,
  RemoteInputSource,
  RemoteRiskLevel,
  RemoteRouteResult,
  RemoteRouteStatus,
  RemoteUserPolicyContext,
} from "./types";

const rateHits = new Map<string, number[]>();
const RATE_WINDOW_MS = 10_000;
const RATE_MAX = 24;

function isRateLimited(commandType: RemoteCommandType): boolean {
  const now = Date.now();
  const list = rateHits.get(commandType) ?? [];
  const fresh = list.filter((t) => now - t < RATE_WINDOW_MS);
  rateHits.set(commandType, fresh);
  return fresh.length >= RATE_MAX;
}

function recordHit(commandType: RemoteCommandType): void {
  const now = Date.now();
  const list = rateHits.get(commandType) ?? [];
  list.push(now);
  rateHits.set(commandType, list);
}

function inferRiskLevel(type: RemoteCommandType): RemoteRiskLevel {
  if (isHighRiskCommandType(type)) return "HIGH";
  return "LOW";
}

function humanLabel(type: RemoteCommandType): string {
  return type
    .replace(/_/g, " ")
    .toLowerCase()
    .replace(/\b\w/g, (c) => c.toUpperCase());
}

export function createRemoteCommand(params: {
  type: RemoteCommandType;
  state: RemoteControlState;
  payload?: Record<string, unknown>;
  inputSource?: RemoteInputSource;
}): RemoteCommand {
  const inputSource = params.inputSource ?? params.state.inputSource;
  const riskLevel = inferRiskLevel(params.type);
  return {
    id:
      typeof crypto !== "undefined" && "randomUUID" in crypto
        ? crypto.randomUUID()
        : `cmd_${Date.now()}_${Math.random().toString(36).slice(2, 9)}`,
    type: params.type,
    label: humanLabel(params.type),
    surface: params.state.surface,
    riskLevel,
    inputSource,
    payload: {
      ...params.state.activeContext,
      ...params.payload,
    },
    requiresConfirmation: riskLevel === "HIGH",
  };
}

export async function routeRemoteCommand(params: {
  command: RemoteCommand;
  state: RemoteControlState;
  userPolicy: RemoteUserPolicyContext;
  handlers: RemoteCommandHandlerMap;
}): Promise<RemoteRouteResult> {
  const { command, state, userPolicy, handlers } = params;

  if (command.riskLevel === "BLOCKED" || command.disabledReason) {
    return {
      status: "blocked",
      reason: command.disabledReason ?? "Command is blocked.",
      command,
    };
  }

  if (isRateLimited(command.type)) {
    return { status: "rate_limited", reason: "Too many commands. Slow down.", command };
  }

  const permission = evaluateRemotePermission(command, userPolicy, state.isLocked);
  if (!permission.allowed) {
    return { status: "blocked", reason: permission.reason, command };
  }

  const mustConfirm =
    permission.requiresConfirmation &&
    (command.riskLevel === "HIGH" || command.riskLevel === "MEDIUM");

  if (mustConfirm) {
    return {
      status: "confirmation_required",
      confirmationCopy: buildConfirmationCopy(command),
      command,
    };
  }

  const handler = handlers[command.type];
  if (!handler) {
    return {
      status: "ignored",
      reason: "No handler registered for this command.",
      command,
    };
  }

  recordHit(command.type);
  await handler(command);
  return { status: "executed", command };
}

/** Same pipeline as {@link routeRemoteCommand} (barrel / external API name). */
export const executeRemoteCommand = routeRemoteCommand;

export async function confirmRemoteCommand(params: {
  command: RemoteCommand;
  handlers: RemoteCommandHandlerMap;
}): Promise<RemoteRouteResult> {
  const handler = params.handlers[params.command.type];
  if (!handler) {
    return {
      status: "ignored",
      reason: "No handler registered for this command.",
      command: params.command,
    };
  }

  recordHit(params.command.type);
  await handler(params.command);
  return { status: "executed", command: params.command };
}

export function commandResultToLogItem(result: RemoteRouteResult): RemoteCommandLogItem {
  const cmd = result.command;
  const at = new Date().toISOString();
  return {
    id: cmd?.id ?? `log_${Date.now()}_${Math.random().toString(36).slice(2, 7)}`,
    commandType: cmd?.type ?? "UNKNOWN",
    label: cmd?.label,
    status: result.status,
    at,
    reason: result.reason,
  };
}

export function buildRemoteCommandLogItem(
  command: RemoteCommand,
  status: RemoteRouteStatus,
  reason?: string
): RemoteCommandLogItem {
  return {
    id: command.id,
    commandType: command.type,
    label: command.label,
    status,
    at: new Date().toISOString(),
    reason,
  };
}

export function clearRemoteRateLimitMemory(): void {
  rateHits.clear();
}

File: integrations/old-source-preservation/ivault-eye-tracking/snapshot/lib/features/remote/remote_router.dart
import 'remote_commands.dart';
import 'remote_policy.dart';
import 'remote_types.dart';

/// Outcome from [routeRemoteCommand].
sealed class RemoteRouteResult {}

final class RemoteRouteExecuted extends RemoteRouteResult {
  RemoteRouteExecuted(this.command);
  final RemoteCommand command;
}

final class RemoteRouteBlocked extends RemoteRouteResult {
  RemoteRouteBlocked(this.reason);
  final String reason;
}

final class RemoteRouteConfirmationRequired extends RemoteRouteResult {
  RemoteRouteConfirmationRequired({
    required this.command,
    required this.confirmationCopy,
  });
  final RemoteCommand command;
  final String confirmationCopy;
}

final class RemoteRouteRateLimited extends RemoteRouteResult {
  RemoteRouteRateLimited(this.reason);
  final String reason;
}

final class RemoteRouteIgnored extends RemoteRouteResult {
  RemoteRouteIgnored(this.reason);
  final String reason;
}

/// Sliding window rate limiter (in-memory MVP).
final class RemoteRateLimiter {
  RemoteRateLimiter({
    this.maxPerWindow = 24,
    this.window = const Duration(seconds: 10),
  });

  final int maxPerWindow;
  final Duration window;
  final Map<String, List<DateTime>> _hits = {};

  bool isLimited(String commandType) {
    final now = DateTime.now();
    final list = _hits.putIfAbsent(commandType, () => []);
    list.removeWhere((t) => now.difference(t) > window);
    return list.length >= maxPerWindow;
  }

  void record(String commandType) {
    final now = DateTime.now();
    _hits.putIfAbsent(commandType, () => []).add(now);
  }

  void reset() => _hits.clear();
}

String buildConfirmationCopy(RemoteCommand command) {
  return 'Confirm "${command.label}" (${command.type})? '
      'Risk: ${command.riskLevel.name}.';
}

/// Single entry point for remote commands (spec §12–§13).
RemoteRouteResult routeRemoteCommand({
  required RemoteCommand command,
  required RemotePolicyContext policy,
  required RemoteRateLimiter rateLimiter,
  required bool remoteLockedFromState,
  required bool strictConfirmations,
}) {
  if (command.riskLevel == RemoteRiskLevel.blocked ||
      command.disabledReason != null) {
    return RemoteRouteBlocked(
      command.disabledReason ?? 'Command is blocked.',
    );
  }

  if (rateLimiter.isLimited(command.type)) {
    return RemoteRouteRateLimited('Too many commands. Slow down.');
  }

  final permission = evaluateRemotePermission(
    command,
    policy,
    remoteLocked: remoteLockedFromState,
  );
  if (!permission.allowed) {
    return RemoteRouteBlocked(permission.reason ?? 'Not allowed.');
  }

  final mustConfirm = permission.requiresConfirmation &&
      (strictConfirmations ||
          command.riskLevel == RemoteRiskLevel.high ||
          kRemoteCommandRegistry[command.type]?.riskLevel ==
              RemoteRiskLevel.high);

  if (mustConfirm) {
    return RemoteRouteConfirmationRequired(
      command: command,
      confirmationCopy: buildConfirmationCopy(command),
    );
  }

  rateLimiter.record(command.type);
  return RemoteRouteExecuted(command);
}
